scipy/signal/_signaltools.py

```diff
@@ -3910,19 +3949,23 @@ def vectorstrength(events, period):
     # we need to know later if period was originally a scalar
     scalarperiod = not period.ndim
 
-    events = np.atleast_2d(events)
-    period = np.atleast_2d(period)
-    if (period <= 0).any():
+    events = xpx.atleast_nd(events, ndim=2, xp=xp)
+    period = xpx.atleast_nd(period, ndim=2, xp=xp)
+    if xp.any(period <= 0):
         raise ValueError('periods must be positive')
 
     # this converts the times to vectors
-    vectors = np.exp(np.dot(2j*np.pi/period.T, events))
+    I2pi = xp.asarray(2j*xp.pi, dtype=xp.complex64
+                        if period.dtype == xp.float32
+                        else xp.complex128)
+    events_ = xp.astype(events, I2pi.dtype) if is_torch(xp) else events
```

Comment:
why special-case on torch here?

Reply:
```
In [2]: torch.eye(2) @ torch.eye(2, dtype=torch.complex64)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[2], line 1
----> 1 torch.eye(2) @ torch.eye(2, dtype=torch.complex64)

RuntimeError: expected m1 and m2 to have the same dtype, but got: float != c10::complex<float>

```

Reply:
that looks like an array-api-compat bug then. [The spec says that](https://data-apis.org/array-api/draft/API_specification/generated/array_api.array.__matmul__.html#api-specification-generated-array-api-array-matmul--page-root)

> The returned array must have a data type determined by [Type Promotion Rules](https://data-apis.org/array-api/draft/API_specification/type_promotion.html#type-promotion).